### PR TITLE
fix(profile): consider potential split db in profile icondirect

### LINKED ISSUE #1997: profile: when split database is implemented, update caching of profile icons (Trac #1997)
_Original ticket http://trac.elgg.org/ticket/1997 on 40214589-09-28 by trac user bios, assigned to unknown._

Elgg version: 1.7

Same as in [http://trac.elgg.org/ticket/1996] database split logic isn`t considered in profile plugin.

At file "/mod/profile/icondirect.php" at line 48:

    if ($mysql_dblink = mysql_connect($CONFIG->dbhost,$CONFIG->dbuser,$CONFIG->dbpass, true)) {

DB split settings isn`t taken into account.

### PR DESCRIPTION
**Pull request by jdalsem**
fixes #1997

### CODE REVIEW

File: mod/profile/icondirect.php [hunk 1]
```diff
@@ -38,9 +38,34 @@
...
+$host = $CONFIG->dbhost;
+$dbuser = $CONFIG->dbuser;
+$dbpass = $CONFIG->dbpass;
+$dbname = $CONFIG->dbname;
+
+// check if split db
+if (isset($CONFIG->db) && isset($CONFIG->db['split']) && $CONFIG->db['split']) {
+	// get random read connection
+	$random_key = array_rand($CONFIG->db['read']);
+	$read_connection = $CONFIG->db['read'][$random_key];
```
Comment (mrclay) [inline mod/profile/icondirect.php]: `$read_connection = (array)$CONFIG->db['read'][$random_key];`

File: mod/profile/icondirect.php [hunk 2]
```diff
@@ -38,9 +38,34 @@
...
 }
 
-$mysql_dblink = @mysql_connect($CONFIG->dbhost, $CONFIG->dbuser, $CONFIG->dbpass, true);
+$host = $CONFIG->dbhost;
+$dbuser = $CONFIG->dbuser;
+$dbpass = $CONFIG->dbpass;
+$dbname = $CONFIG->dbname;
+
+// check if split db
+if (isset($CONFIG->db) && isset($CONFIG->db['split']) && $CONFIG->db['split']) {
```
Comment (mrclay) [inline mod/profile/icondirect.php]: I think that can just be: `if (!empty($CONFIG->db['split'])) {`

**Comment by Srokap** [comment]
I so much don't like this approach. We should just load autoloader and instantiate just `Elgg_Database` instead of duplicating the logic. This PR does ignore format of the single read configuration ("old style single connection"). That does complicate things without good reason, but nevertheless that's still difference against current core implementation that should cause failures for old configurations.

Right now we assume that it's only possible by loading core and plugins, but that's not hte case anymore. This implementation disregards logging, so if we did it as well in OO approach, and provided some logger stripped of the hooks triggering, we could reuse all the connection establishing logic.

I don't strictly veto this solution, but I do plan to change it in very near future. I'd like all current handlers to go through ElggApplication and only differ in that if it's run that bootstraps the core without plugins or with them. In this case we don't have much reason to be afraid of autoloading itself being particularly costly and Elgg_Database is pretty leightweight (as it should!) , so we don't need to repeat the code and maintain the same performance.

For now I'd ask to use Elgg\Database\Config to fetch DB credentials at least. Something like (not tested):

```php
require_once(COMPOSER_AUTOLOADER_PATH);

global $CONFIG;
$conf = new \Elgg\Database\Config($CONFIG);


if ($conf->isDatabaseSplit()) {
	$read_connection = $conf->getConnectionConfig('read');
} else {
	$read_connection = $conf->getConnectionConfig('readwrite');
}

```
Would make it way more DRY.

**Comment by jdalsem** [comment]
updated as per suggestion of @Srokap (which is much better :)

**Comment by Srokap** [comment]
Awesome! visually LGTM, but don't have time to test right now.

**Pull request merged**